edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/SubscriptionChangeHandler.cs

```diff
@@ -86,45 +91,211 @@ public async Task<bool> HandleAsync(MqttPublishInfo publishInfo)
+                                Events.FailedToChangeSubscriptionState(e, subscriptionPattern.Subscription.ToString(), identity.Id);
                             }
                         }
                         else
                         {
-                            Events.CouldNotObtainListener(subscriptionPattern.Subscrition.ToString(), identity.Id);
+                            Events.CouldNotObtainListenerForSubscription(subscriptionPattern.Subscription.ToString(), identity.Id);
                         }
                     }
                 }
             }
 
-            return true;
+            // At this point we have a collection about 'affectedClients'. Those are the clients subscribed to something.
+            // What we don't know are the unsubscribtions. That is because the broker sends only the current subscription list,
+            // so when a client unsubscribes, that is only a shorter list of subscriptions sent.
+            // The information we want to gain is that what the unsubscriptions are, and the answer is that what we have not seen
+            // previously as subscribed, that is unsubscribed. E.g. if a client subscribed to 'methods' only, then we say that
+            // it unsubscribed from twin responses, c2d, m2m, etc. The underlaying layers take care of the optimization doing
+            // nothing when unsubscribed twice from a topic.
+            // As a first step we need all nested connections, because it is possible that the subscribtion list above did not
+            // cover all clients. It can happen when a client had only a single subscription before, but now it unsubscribed.
+            // In that case that client will not be contained by the subscription event sent by the broker, because it has no
+            // subscription at all.
+            var nestConnections = await this.connectionRegistry.GetNestedConnectionsAsync();
+
+            foreach (var connection in nestConnections)
+            {
+                // for a given client we are going to unsubscribe from everything that was not marked as subscribed
+                // from the broker subscription event. Every subscription marked in the broker event is stored in the
+                // 'affectedClients' collection built previously.
+                var subscriptions = default(HashSet<DeviceSubscription>);
+                if (!affectedClients.TryGetValue(connection, out subscriptions))
+                {
+                    subscriptions = new HashSet<DeviceSubscription>();
+                }
+
+                var maybeListener = await this.connectionRegistry.GetDeviceListenerAsync(connection, false);
+
+                foreach (var subscription in this.allSubscriptions)
+                {
+                    if (!subscriptions.Contains(subscription))
+                    {
+                        if (maybeListener.HasValue)
```

Comment:
check listener exists before iterating

Reply:
It needs to print an error message for every subscription, so if there is no listener, and there is C2D and Methods subscription, we would see an error for both C2D and Methods. 
I can do something like this
```
                foreach (var subscription in this.allSubscriptions)
                {
                    if (!subscriptions.Contains(subscription))
                    {
                        await maybeListener.Match(
                            async listener => await AddOrRemoveSubscription(listener, false, subscription),
                            () =>
                            {
                                Events.CouldNotObtainListenerForSubscription(subscription.ToString(), connection.Id);
                                return Task.CompletedTask;
                            });
                    }
                }
```

which looks strange because of the Task.CompletedTask

Reply:
I've changed Events.CouldNotObtainListenerForSubscription() to return Task.Completed, so now the code can be 

```
foreach (var subscription in this.allSubscriptions)
                {
                    if (!subscriptions.Contains(subscription))
                    {
                        await maybeListener.Match(
                            listener => AddOrRemoveSubscription(listener, false, subscription),
                            () => Events.CouldNotObtainListenerForSubscription(subscription.ToString(), connection.Id));
                    }
                }
```